### PR TITLE
Conforming to new CoTeDe config pattern

### PR DESCRIPTION
**Pull request by castelao**
CoTeDe configuration format changed requiring to conform cotede_test.py

@BillMills & @s-good , I released a new patch for CoTeDe and I believe that I fixed everything in the cotede_test.py for AutoQC. I didn't manage to build my Docker so I could run a full test. If it needs more changes, I might take a few days to be able to respond again.

Cheers,

### CODE REVIEW

**Comment by castelao** [comment]
I forgot to mention that this was related to #258

File: cotede_qc/cotede_test.py [hunk 1]
```diff
@@ -43,26 +43,26 @@ def get_qc(p, config, test):
...
                     # Load from TEMP,
                     try:
-                        cfg = {var: {test: cfg[var][test]}}
+                        cfg = {'variables': {'sea_water_temperature': {test: cfg['variables']['sea_water_temperature'][test]}}}
                     # otherwise load it from main.
                     except:
                         # The dummy configuration ensures that the results from
                         # 'main' is copied into the results for var.
-                        cfg = {'main': {test: cfg['main'][test]}, var: {'dummy': None}}
+                        cfg = {'common': {test: cfg['common'][test]}, var: {'dummy': None}}
```
Comment (BillMills) [inline cotede_qc/cotede_test.py:65]: The `var: {'dummy': None}` errors out when the config processing reaches `cotede/utils/config.py  guess_procedure`, complaining that there's no `dummy` in the catalog. I'm not clear on the original intention of this particular line in AutoQC - the first `common` key is what I would expect for the config when `test` is `valid_datetime`, `valid_geolocation`, or `location_at_sea`. What is the purpose of the dummy object? Tests pass if I just add `dummy: None` to the catalog in `guess_procedure`, but again this seems like a kludge.

Possibly related, the catalog in CoTeDe's `guess_procedure` has an entry for `valid_geolocation` but not `location_at_sea` or `valid_datetime`. @castelao is this the intention?

**Comment by castelao** [comment]
Give me a couple of days and I'll get back to this. Thanks!

**Comment by BillMills** [comment]
@castelao yes of course, please take as long as you need - never any hurry here, we appreciate any time you have given how busy we know you are.

I created a PR https://github.com/castelao/AutoQC/pull/1 against this branch with some suggestions on what I had to amend here to get most tests not erroring out. LMKWYT whenever you have time. Thanks very much!

**Comment by BillMills** [comment]
@castelao after including these changes and my guesses at amendments in castelao#1, there's one more error detected in AutoQC's CoTeDe tests, in `qctests/CoTeDe_GTSPP_profile_envelop.py`, which I can trace down to how this new config object is getting consumed in CoTeDe's `cotede/qctests/profile_envelop.py`. 

Specifically, the use of `cfg['layers']` at https://github.com/castelao/CoTeDe/blob/master/cotede/qctests/profile_envelop.py#L54 and https://github.com/castelao/CoTeDe/blob/master/cotede/qctests/profile_envelop.py#L29 doesn't match the current configuration object structure, which I see to be something like:

```
{'layers': OrderedDict([('layers', [['> 0', '<= 25', -2, 37], ['> 25', '<= 50', -2, 36], ['> 50', '<= 100', -2, 36], ['> 100', '<= 150', -2, 34], ['> 150', '<= 200', -2, 33], ['> 200', '<= 300', -2, 29], ['> 300', '<= 400', -2, 27], ['> 400', '<= 1100', -2, 27], ['> 1100', '<= 3000', -1.5, 18], ['> 3000', '<= 5500', -1.5, 7], ['> 5500', '<= 12000', -1.5, 4]]), ('procedure', 'ProfileEnvelop')]), 'procedure': 'ProfileEnvelop'}
```

Notice the double `layers` keys; if in `cotede/qctests/profile_envelop.py` I change `cfg['layers']` to `cfg['layers']['layers']`, the test at least runs.

With this as in all my comments here, I can track down where things are blowing up, but I'm not enough of a CoTeDe expert to insist on what the right course of action is here - relying on your expertise to sort out this config object construction.

File: cotede_qc/cotede_test.py [hunk 2]
```diff
@@ -43,26 +43,26 @@ def get_qc(p, config, test):
...
-                            del(cfg[v])
+                    for v in cfg['variables'].keys():
+                        if v != 'sea_water_temperature':
+                            del(cfg['variables'][v])
                 # If is a specific test,
                 elif test != config:
                     # Load from TEMP,
                     try:
-                        cfg = {var: {test: cfg[var][test]}}
+                        cfg = {'variables': {'sea_water_temperature': {test: cfg['variables']['sea_water_temperature'][test]}}}
```
[the remark this answers is not in the document]
Comment (castelao) [reply cotede_qc/cotede_test.py]: Nice! I'm sorry that I couldn't respond to this before. And worse, I forgot about this suggestion here. Yep, I just submitted a new PR with the same changes.

**Comment by castelao** [comment]
Hi @BillMills . I'm sorry, I jump straight into my local repo and didn't check here before. I forgot that you had made some corrections. I believe that your corrections were all that we needed to run with the lastest CoTeDe. Oh well, at least I wasted my own time.

I'm not sure why it is failing the ci. Looks like it is running an older version of CoTeDe?

**Comment by castelao** [comment]
There were a couple of changes here:

 - I added a few new QC configs that might be interesting in the ROC plot in the paper. It changes the Anomaly Detection threshold, so it should transition between minimizing false positives or false negatives.
 - The qc_returned now is the 'overall'. Since all other flags are removed just before applying the QC, simple tests will be the same result, but for compound tests it will return the overall combined result. For instance, 'gtspp' will return in the overall combined result for the full GTSPP procedure.

**Comment by BillMills** [comment]
Thanks very much for this, @castelao - I think we're almost there. You can run the test suite in the dockerized environment by doing:

```
docker container run -it iquod/autoqc:PR259-debug bash
nosetests tests/CoTeDe_checks_validation.py 
```

When I do so, I see 5 fails and 2 errors. I'm not concerned at this time about the fails - some results will change due to the update in the oceansdb version and parameters, which will look like failures in some tests. The errors are more significant, but I have suggestions for both:

## Argo density inversion

Reports error:

```
======================================================================
ERROR: Make sure CoTeDe Argo density inversion is working as expected.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/AutoQC/tests/CoTeDe_checks_validation.py", line 33, in test_CoTeDe_Argo_density_inversion
    qc = qctests.CoTeDe_Argo_density_inversion.test(padi, None)
  File "/AutoQC/qctests/CoTeDe_Argo_density_inversion.py", line 10, in test
    qc = get_qc(p, config, testname)
  File "/AutoQC/cotede_qc/cotede_test.py", line 71, in get_qc
    pqc = ProfileQC(inputs, cfg=cfg)
  File "/usr/local/lib/python3.6/dist-packages/cotede/qc.py", line 100, in __init__
    self.evaluate(v, self.cfg['variables'][c])
  File "/usr/local/lib/python3.6/dist-packages/cotede/qc.py", line 332, in evaluate
    self.flags[v]['overall'] = combined_flag(self.flags[v])
  File "/usr/local/lib/python3.6/dist-packages/cotede/misc.py", line 25, in combined_flag
    output = np.asanyarray(flags[criteria[0]])
IndexError: list index out of range
```

Digging around in the code, it appears that in `cotede/qc.py` in the `evaluate` function, there's no longer a condition describing what to do with tests labeled `'density_inversion'`. A condition to handle this was present in CoTeDe 0.19.2 (the version we were using previously), but was removed here: https://github.com/castelao/CoTeDe/commit/6f26a23974201df25dc15001dbf1ac6885e7cbe0#. What's the correct way to call the Argo density inversion now?

## CoTeDe profile_envelop

Reports error:

```
======================================================================
ERROR: Make sure CoTeDe profile_envelop is working as expected.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/AutoQC/tests/CoTeDe_checks_validation.py", line 129, in test_CoTeDe_GTSPP_profile_envelop
    qc = qctests.CoTeDe_GTSPP_profile_envelop.test(p, None)
  File "/AutoQC/qctests/CoTeDe_GTSPP_profile_envelop.py", line 10, in test
    qc = get_qc(p, config, testname)
  File "/AutoQC/cotede_qc/cotede_test.py", line 71, in get_qc
    pqc = ProfileQC(inputs, cfg=cfg)
  File "/usr/local/lib/python3.6/dist-packages/cotede/qc.py", line 100, in __init__
    self.evaluate(v, self.cfg['variables'][c])
  File "/usr/local/lib/python3.6/dist-packages/cotede/qc.py", line 249, in evaluate
    y = Procedure(self.input, v, cfg[criterion], autoflag=True)
  File "/usr/local/lib/python3.6/dist-packages/cotede/qctests/qctests.py", line 83, in __init__
    super().__init__(data, cfg, autoflag)
  File "/usr/local/lib/python3.6/dist-packages/cotede/qctests/qctests.py", line 50, in __init__
    self.test()
  File "/usr/local/lib/python3.6/dist-packages/cotede/qctests/profile_envelop.py", line 55, in test
    ind = np.nonzero(eval("(z %s) & (z %s)" % (layer[0], layer[1])))[0]
  File "<string>", line 1
    (z l) & (z a)
       ^
SyntaxError: invalid syntax
```

As I mention above (https://github.com/IQuOD/AutoQC/pull/259#issuecomment-687901078), looking at `cotede/qctests/profile_envelop.py` there appears to be a key missing; all references to `cfg['layers']` seem to need to be `cfg['layers']['layers']`, at which point the test passes.

Regarding the CI: it's failing right now because it's running in a fixed containerized environment, rather than updating CoTeDe and oceansdb. The reason for this is so that we can always reproduce the environment that CI passes in by using the same container, rather than relying on a live build that may be difficult to recreate after the fact. Anyway, this doesn't matter too much right now - I'll update the CI testing environment when we get the above tests passing.

**Comment by BillMills** [comment]
All these commits included in #261, plus corrections to get tests passing. See comments and discussion there, and thanks again @castelao for this!

**Pull request merged**